### PR TITLE
[SYCL][NATIVECPU] Enable source-based code coverage in Native CPU

### PR DESCRIPTION
**Pull request by uwedolinsky**
Supports [clang's source-based code coverage](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html) to enable code coverage testing of SYCL applications via the Native CPU SYCL target.

Clang's `-fprofile-instr-generate -fcoverage-mapping` options can now be used with the `native_cpu` SYCL target to compile/instrument host and device code, enabling 'llvm-cov' to render a coverage report after running the SYCL application (see also updated documentation in this PR). 

Subsequent PRs will enable in NativeCPU more of the currently unsupported options for device compilation, also for performance profiling.

**Details and explanations for the changes in this PR:**
This PR tests coverage options on the existing NativeCPU vector-add test by adding an additional invocation with previously disabled options `-fprofile-instr-generate -fcoverage-mapping -mllvm -system-headers-coverage`. Enabling these options on device code caused an [assert in the upstream clang profiling code generation tools](https://github.com/intel/llvm/blob/b023d407862bd853ba5881c34985f99d039d856c/clang/lib/CodeGen/CoverageMappingGen.cpp#L960)  due to the invalid source location on the AST for the implicitly generated kernel body, specifically the compound statement containing the kernel body. This PR honors this upstream clang assert by replacing the invalid source location in the compound statement with the source location of the kernel body. Using this now valid source location maintains the location (of the kernel caller function) currently tested by [non-upstream-llvm lit test `CodeGenSYCL/debug-info-srcpos-kernel.cpp`](https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/debug-info-srcpos-kernel.cpp), but exposed an issue that led to a change in behavior in [non-llvm-upstream lit test `SemaSYCL/kernel-arg-opt-report.cpp`](https://github.com/intel/llvm/blob/sycl/clang/test/SemaSYCL/kernel-arg-opt-report.cpp), which was due to the previously invalid source location causing the compiler to skip code to set the current location. To restore the original behavior of this test (checking for the location of the kernel functor, as opposed to the kernel caller function) this PR temporarily (and only for the purpose of generating the report) sets the current location to the location of the kernel argument using the upstream clang utility [clang::CodeGen::ApplyDebugLocation](https://github.com/intel/llvm/blob/b023d407862bd853ba5881c34985f99d039d856c/clang/lib/CodeGen/CGDebugInfo.h#L860).

### CODE REVIEW

**Comment by uwedolinsky** [comment]
@intel/dpcpp-tools-reviewers  @intel/dpcpp-cfe-reviewers @intel/dpcpp-clang-driver-reviewers please could you review? Thank you!

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp [hunk 1]
```diff
@@ -244,7 +244,8 @@ class DependencyGraph {
         Graph[UFunc].insert(V);
       } else if (isa<const Constant>(U)) {
         if (const auto *GV = dyn_cast<const GlobalVariable>(U))
-          Graph[GV].insert(V);
+          if (Graph[GV].insert(V).second == false)
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:247]: can you explain this change? it's unclear why its needed from the title and description. thanks

**Review by mdtoguchi — APPROVED** [review]
Just some small comment updates - otherwise looks OK for driver

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp (hunk 1 above)
Comment (uwedolinsky) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:247]: You are right, the compiler invocation that needs this change is not tested by this PR, so I'll revert this change for now. I'll open a subsequent separate PR for this. Thank you.

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 2]
```diff
@@ -1786,6 +1786,9 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
     if (SyclOptReport.HasOptReportInfo(FD)) {
       llvm::OptimizationRemarkEmitter ORE(Fn);
       for (auto ORI : llvm::enumerate(SyclOptReport.GetInfo(FD))) {
+        // Temporarily apply arg location to ensure SourceLocToDebugLoc
+        // picks up the expected file.
+        ApplyDebugLocation tmp_raii(*this, ORI.value().KernelArgLoc);
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1791]: Please use Pascal Case for naming variables. `TempRAII  is an option.

It also isn't clear to me where this is used.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -3602,8 +3602,12 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     BodyStmts.insert(BodyStmts.end(), FinalizeStmts.begin(),
                      FinalizeStmts.end());
 
+    const auto *LDcl = NewBody;
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:3605]: Why can't you just use `NewBody` instead of creating a new Stmt?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -3602,8 +3602,12 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
...
+    SourceLocation LL = LDcl ? LDcl->getBeginLoc() : SourceLocation();
+    SourceLocation LR = LDcl ? LDcl->getEndLoc() : SourceLocation();
+
     return CompoundStmt::Create(SemaSYCLRef.getASTContext(), BodyStmts,
-                                FPOptionsOverride(), {}, {});
+                                FPOptionsOverride(), LL, LR);
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:3610]: What is this actually doing? This is supposed to correspond to location in source code IIUC which of course doesn't exist for compiler generated code. So what location is being picked up here?

**Review by elizabethandrews — COMMENTED** [review]
It isn't clear to me what the FE changes are doing in this patch. Please add a FE test if there is functionality that can be tested with these changes. It looks like you are only testing driver changes here. If that is the only functionality that can be tested at the moment, please remove FE changes from this PR and add them when it can be tested.

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp (hunk 1 above)
Comment (uwedolinsky) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp]: This is now reverted

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 2 above)
Comment (uwedolinsky) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: I've changed the variable name, hope it's ok. 

This variable isn't (directly) used by the subsequent code in this function - it's just a RAII variable to temporarily set the argument location to maintain the current behavior of this [non-llvm-upstream lit test](https://github.com/intel/llvm/blob/sycl/clang/test/SemaSYCL/kernel-arg-opt-report.cpp). I have updated the PR description with more details. Please let me know if more clarification is needed. Thanks.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (uwedolinsky) [reply clang/lib/Sema/SemaSYCL.cpp:3609]: This is picking up the source location from the kernel body - so the compound statement uses the same location as the kernel body. You are right, this is a compiler-generated function, but it still needs a valid source location to ensure any tools processing the clang AST and requiring that source location to be valid to succeed. Other compiler-generated function like constructors also have valid locations even though they don't exist in source code.

**Comment by uwedolinsky** [comment]
> It isn't clear to me what the FE changes are doing in this patch. Please add a FE test if there is functionality that can be tested with these changes. It looks like you are only testing driver changes here. If that is the only functionality that can be tested at the moment, please remove FE changes from this PR and add them when it can be tested.

@elizabethandrews I have updated the PR description with more details and explanations about the changes. Please let me know if anything needs further clarification. Thanks

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:3609]: I think this sounds reasonable but I would like @tahonermann to also weigh in here. IIRC the work we re currently implementing for upstream clang (which will replace a lot of this code eventually) does not set SourceLocations as well.

Comment (uwedolinsky) [reply clang/lib/Sema/SemaSYCL.cpp:3609]: @elizabethandrews @tahonermann do you have updates on this? Thank you

Comment (tahonermann) [reply clang/lib/Sema/SemaSYCL.cpp:3609]: Sorry, I somehow missed @elizabethandrews's earlier request for me to weigh in on this.

I think this is reasonable. I'm not sure there is clear guidance for cases like this where a function definition is composed from both generated and written code.

There might be cases in the new upstream implementation that will need to have source locations added. I think I generally opted for generated code to have builtin/unknown source locations.

**Comment by elizabethandrews** [comment]
> > It isn't clear to me what the FE changes are doing in this patch. Please add a FE test if there is functionality that can be tested with these changes. It looks like you are only testing driver changes here. If that is the only functionality that can be tested at the moment, please remove FE changes from this PR and add them when it can be tested.
> 
> @elizabethandrews I have updated the PR description with more details and explanations about the changes, which are all covered by existing tests. Please let me know if anything needs further clarification. Thanks

@uwedolinsky FE changes need FE tests. The tests you added are to test driver changes and an E2E test verifying FE changes. I do not think this is sufficient. IIUC you patch should have changed source locations in AST. This should be tested via a FE AST test.

**Comment by uwedolinsky** [comment]
> @uwedolinsky FE changes need FE tests. The tests you added are to test driver changes and an E2E test verifying FE changes. I do not think this is sufficient. IIUC you patch should have changed source locations in AST. This should be tested via a FE AST test.

@elizabethandrews  I've added a new AST test to check that the compound statement of the kernel body has a valid source location. Is that acceptable?

**Review by elizabethandrews — APPROVED** [review]

**Comment by elizabethandrews** [comment]
> > @uwedolinsky FE changes need FE tests. The tests you added are to test driver changes and an E2E test verifying FE changes. I do not think this is sufficient. IIUC you patch should have changed source locations in AST. This should be tested via a FE AST test.
> 
> @elizabethandrews I've added a new AST test to check that the compound statement of the kernel body has a valid source location. Is that acceptable?

Yes. Thank you

**Review by PietroGhg — APPROVED** [review]

**Comment by uwedolinsky** [comment]
@intel/dpcpp-tools-reviewers please could you review? Thank you

**Comment by uwedolinsky** [comment]
@intel/llvm-gatekeepers this PR looks ready to merge. The approval from @intel/dpcpp-tools-reviewers seems no longer required because I [reverted a modifcation](https://github.com/intel/llvm/pull/15073/commits/6b72582ace1617a8bb8028cd08ffb5088a500502) to a file owned by this group. Please could you double-check? Thank you

**Pull request merged**